saq/queue/postgres.py

```diff
@@ -0,0 +1,521 @@
+
+        return {
+            "workers": {},
+            "name": self.name,
+            "queued": queued,
+            "active": active,
+            "scheduled": incomplete - queued - active,
+            "jobs": [],
+        }
+
+    async def stats(self, ttl: int = 60) -> QueueStats:
+        return {
+            "complete": self.complete,
+            "failed": self.failed,
+            "retried": self.retried,
+            "aborted": self.aborted,
+            "uptime": now() - self.started,
+        }
+
+    async def count(self, kind: CountKind) -> int:
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            if kind == "queued":
+                await cursor.execute(
+                    f"""
+                    SELECT count(*) FROM {self.jobs_table}
+                    WHERE status = 'queued'
+                      AND queue = %(queue)s
+                      AND %(now)s >= scheduled
+                    """,
+                    {"queue": self.name, "now": math.ceil(seconds(now()))},
+                )
+            elif kind == "active":
+                await cursor.execute(
+                    f"""
+                    SELECT count(*) FROM {self.jobs_table}
+                    WHERE status = 'active'
+                      AND queue = %(queue)s
+                    """,
+                    {"queue": self.name},
+                )
+            elif kind == "incomplete":
+                await cursor.execute(
+                    f"""
+                    SELECT count(*) FROM {self.jobs_table}
+                    WHERE status IN ('new', 'deferred', 'queued', 'active')
+                      AND queue = %(queue)s
+                    """,
+                    {"queue": self.name},
+                )
+            else:
+                raise ValueError("Can't count unknown type {kind}")
+
+            result = await cursor.fetchone()
+            assert result
+            return result[0]
+
+    async def schedule(self, lock: int = 1) -> t.Any: ...
+
+    async def sweep(self, lock: int = 60, abort: float = 5.0) -> list[t.Any]:
+        """Delete jobs past their TTL from the jobs table"""
+        async with self.pool.connection() as conn, conn.cursor() as cursor:
+            await cursor.execute(
+                f"""
+                DELETE FROM {self.jobs_table}
+                WHERE status IN ('aborted', 'complete', 'failed')
+                  AND %(now)s >= ttl
+                RETURNING key
+                """,
+                {"now": math.ceil(seconds(now()))},
+            )
+            await cursor.execute(
+                f"""
+                SELECT job
+                FROM {self.jobs_table} LEFT OUTER JOIN pg_locks ON lock_id = objid
+                WHERE status = 'active'
+                  AND objid IS NULL
+                """
+            )
+            results = await cursor.fetchall()
+        swept = []
+        for result in results:
```

Comment:
what happens if the sweeper gets killed after the delete from jobs table but before retry / abort?

Reply:
The delete above is for clearing out jobs with terminal statuses past their ttl, so it's not related to retry/abort. However, all of sweep now takes place in a transaction, so everything should be rolled back if the sweeper gets killed before it finishes.